sdks/python/apache_beam/runners/dataflow/dataflow_metrics.py

```diff
@@ -30,25 +30,60 @@
 from apache_beam.metrics.metricbase import MetricName
 
 
-# TODO(pabloem)(JIRA-1381) Implement this once metrics are queriable from
-# dataflow service
 class DataflowMetrics(MetricResults):
   """Implementation of MetricResults class for the Dataflow runner."""
 
-  def __init__(self, dataflow_client=None, job_result=None):
+  def __init__(self, dataflow_client=None, job_result=None, job_graph=None):
     """Initialize the Dataflow metrics object.
 
     Args:
       dataflow_client: apiclient.DataflowApplicationClient to interact with the
         dataflow service.
       job_result: DataflowPipelineResult with the state and id information of
-        the job
+        the job.
+      job_graph: apiclient.Job instance to be able to translate between internal
+        step names (e.g. "s2"), and user step names (e.g. "split").
     """
     super(DataflowMetrics, self).__init__()
     self._dataflow_client = dataflow_client
     self.job_result = job_result
     self._queried_after_termination = False
     self._cached_metrics = None
+    self._job_graph = job_graph
+
+  def _translate_step_name(self, internal_name):
+    """Translate between internal step names (e.g. "s1") and user step names."""
+    if not self._job_graph:
+      raise ValueError('Could not translate the internal step name.')
+
+    try:
+      [step] = [step
+                for step in self._job_graph.proto.steps
+                if step.name == internal_name]
+      [user_step_name] = [prop.value.string_value
+                          for prop in step.properties.additionalProperties
+                          if prop.key == 'user_name']
+    except ValueError:
+      raise ValueError('Could not translate the internal step name.')
+    return user_step_name
+
+  def _get_metric_key(self, metric):
+    """Populate the MetricKey object for a queried metric result."""
+    try:
+      [step] = [prop.value
+                for prop in metric.name.context.additionalProperties
+                if prop.key == 'step']
+      step = self._translate_step_name(step)
+      [namespace] = [prop.value
+                     for prop in metric.name.context.additionalProperties
+                     if prop.key == 'namespace']
+      name = metric.name.name
+    except ValueError:
+      # An unstructured metric name is "step/namespace/name", but step names
+      # can (and often do) contain slashes. Must only split on the right-most
+      # two slashes, to preserve the full step name.
+      [step, namespace, name] = metric.name.name.rsplit('/', 2)
```

Comment:
The possibilities are:
1. Unable to unpack `[step]` in line 73. This means that the counter does not have a `"step"` property. This is never the case for metrics.
2. Unable to translate the step name. This happens only if the job graph is improperly formatted (seems really unlikely); or if the step name is not the internal step name (this happens for unstructured names only - so we're fine there).
3. Unable to unpack `[namespace]` in line 77. This would mean that the step does not have a `"namespace"` property. (e.g. is not a structured metric).

There could be problems if we have Aggregators, but they have been removed from this SDK, so this should be safe as long as the only metrics with `"user"` origin are coming from the metrics API. If there are non-metrics-api counters that have `"user"` origin, then there may be problems.